### PR TITLE
No-streaming Mask

### PR DESCRIPTION
**Pull request by Olllom**
This PR changes the way masks are applied.
It allows boundary conditions to define `make_no_stream_mask` and `make_no_collision_mask` functions.

@MartinKliemank

### CODE REVIEW

File: lettuce/simulation.py [hunk 1]
```diff
@@ -125,10 +133,10 @@ def initialize_f_neq(self):
...
             S = torch.cat([S, grad_u2])
 
-        Pi_1 = 1.0 * self.flow.units.relaxation_parameter_lu * rho  * S / self.lattice.cs**2
-        Q = torch.einsum('ia,ib->iab',self.lattice.e,self.lattice.e)-torch.eye(self.lattice.D,device=self.lattice.device)*self.lattice.cs**2
-        Pi_1_Q = torch.einsum('ab...,iab->i...', Pi_1, Q)
-        fneq = torch.einsum('i,i...->i...',self.lattice.w,Pi_1_Q)
+        Pi_1 = 1.0 * self.flow.units.relaxation_parameter_lu * rho * S / self.lattice.cs**2
+        Q = torch.einsum('ia,ib->iab', [self.lattice.e, self.lattice.e]) - torch.eye(self.lattice.D, device=self.lattice.device, dtype=self.lattice.dtype)*self.lattice.cs**2
+        Pi_1_Q = self.lattice.einsum('ab,iab->i', [Pi_1, Q])
+        fneq = self.lattice.einsum('i,i->i', [self.lattice.w, Pi_1_Q])
```
Comment (Olllom) [inline lettuce/simulation.py:139]: don't mind this. this has nothing to do with the boundary conditions

File: lettuce/boundary.py [hunk 2]
```diff
@@ -93,3 +108,14 @@ def __call__(self, f):
...
+    def make_no_stream_mask(self, f_shape):
+        no_stream_mask = torch.zeros(size=f_shape, dtype=torch.bool, device=self.lattice.device)
+        no_stream_mask[[np.array(self.lattice.stencil.opposite)[self.velocities]] + self.index] = 1
+        return no_stream_mask
+
+    # not sure about this. it probably makes sense to apply collisions here so that f[0] can adapt to the flow
+    #def make_no_collision_mask(self, f_shape):
+    #    no_collision_mask = torch.zeros(size=f_shape[1:], dtype=torch.bool, device=self.lattice.device)
+    #    no_collision_mask[self.index] = 1
+    #    return no_collision_mask
```
Comment (MartinKliemank) [inline lettuce/boundary.py:121]: As far as I understand collisions should be applied here since the equation explicitly includes the post collision distributions

File: lettuce/boundary.py [hunk 3]
```diff
@@ -85,6 +99,7 @@ def __init__(self, lattice, direction):
             self.w = self.lattice.w[self.velocities]
 
     def __call__(self, f):
+        # overwrite streamed fs with stored fs
```
Comment (MartinKliemank) [inline lettuce/boundary.py:102]: Why call them stored fs? We calculate them as a response to the current flow situation on the boundary, don't we?

File: lettuce/simulation.py [hunk 4]
```diff
@@ -44,13 +44,21 @@ def __init__(self, flow, lattice, collision, streaming):
 
         self.reporters = []
 
-        # Define a mask, where the collision shall not be applied
+        # Define a mask, where the collision or streaming shall not be applied
+        # Define a mask where streaming is not applied
```
Comment (MartinKliemank) [inline lettuce/simulation.py:48]: The second comment seems redundant?

File: lettuce/streaming.py [hunk 5]
```diff
@@ -7,15 +7,38 @@
...
+        self._no_stream_mask = mask
 
     def __call__(self, f):
-        for i in range(self.lattice.Q):
-            f[i] = torch.roll(f[i], shifts=tuple(self.lattice.stencil.e[i]), dims=tuple(np.arange(self.lattice.D)))
+        for i in range(1, self.lattice.Q):
+            if self.no_stream_mask is None:
+                f[i] = self._stream(f, i)
+            else:
+                new_fi = self._stream(f, i)
```
Comment (MartinKliemank) [inline lettuce/streaming.py:35]: Doesn't this use twice the memory? (for the one i?)
I had tried putting the roll directly into the where, though I'm not sure if that is any better:

`f[i] = torch.where(self.no_streaming_mask[i], f[i], torch.roll(f[i], shifts=tuple(self.lattice.stencil.e[i]), dims=tuple(np.arange(self.lattice.D))))`

**Review by MartinKliemank — COMMENTED** [review]
Seems good to me. See questions and suggestions below.

File: lettuce/boundary.py (hunk 2 above)
Comment (Olllom) [reply lettuce/boundary.py:121]: I think it refers to the post-collision distributions coming from inside the domain. The other ones are only relevant for computing u_w. I don't think the Krüger book is specific about how to compute it.


But to me it makes sense to collide here. I think the effect will be small anyway. But if you want to do some more digging, you are welcome.

File: lettuce/boundary.py (hunk 3 above)
Comment (Olllom) [reply lettuce/boundary.py:102]: ah, that's left over from intermediate code
will be deleted

File: lettuce/simulation.py (hunk 4 above)
Comment (Olllom) [reply lettuce/simulation.py:48]: yes

File: lettuce/streaming.py (hunk 5 above)
Comment (Olllom) [reply lettuce/streaming.py:35]: not sure. didn't test. I think this behaves like the previous simpler code but you can try if it makes a difference.

Anyway, you need to make some kind of copy. Otherwise you cannot distinguish between the rolled fi and the original.

**Comment by MartinKliemank** [comment]
Also i think the Equilibrium Boundary (however good or bad it might be) needs a no_streaming_mask aswell, since it produces its own incoming populations?
